### PR TITLE
[BESU-182] Add tests to FilterParameter to demonstrate loading topics from JSON

### PR DESCRIPTION
**Pull request by rolfyone**
Demonstrate that filter topics in JSON format consistent with eth_newFilter RPC requests
will be correctly loaded into the FilterParameter topics.

Previous tests were loading from an internal structure, which makes it difficult to see
coverage of scenarios.

 - [] - matches anything

 - [A] matches A in the first position, and anything after

 - [A, B] - A in first position, and B in second position, then anything after

 - [null, B] - matches anything in first position, and B in second position, and anything after.

 - [[A,B],[A,B]] - Matches A OR B in first position, and A OR B in second position, and anything after.

Jira: [BESU-182]
Signed-off-by: Paul Harris <paul.harris@consensys.net>

### CODE REVIEW

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java [hunk 1]
```diff
@@ -224,6 +243,91 @@ public void emptyListAsSubTopicDecodesCorrectly() throws JsonProcessingException
             emptyList());
   }
 
+  @Test
+  public void emptyListOfTopicsDecodesCorrectly() throws java.io.IOException {
```
Comment (rain-on) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: nit: the naming is as good as the rest of Besu tests, but I always worry when I see "correctly" in tests, what is correctly? This is effectively going to be our requirements doc, so it needs to be explicit about the expected outcome.
eg. emptyListTopicStringDecodesToZeroLengthTopicList ... ok ... that's not great either. :/

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java [hunk 2]
```diff
@@ -224,6 +243,91 @@ public void emptyListAsSubTopicDecodesCorrectly() throws JsonProcessingException
...
+    final FilterParameter filterParameter = readJsonAsFilterParameter("{\"topics\":[]}");
+
+    assertThat(filterParameter.getTopics().size()).isEqualTo(0);
+  }
+
+  @Test
+  public void emptyListOfTopicsNestedDecodesCorrectly() throws java.io.IOException {
+    final FilterParameter filterParameter = readJsonAsFilterParameter("{\"topics\":[[]]}");
+
+    assertThat(filterParameter.getTopics()).containsExactly(List.of());
```
Comment (rain-on) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: nit: _personally_ would prefer to see "emptyList()" ... but maybe that's arbitrary :)

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java [hunk 3]
```diff
@@ -224,6 +243,91 @@ public void emptyListAsSubTopicDecodesCorrectly() throws JsonProcessingException
...
+
+    assertThat(filterParameter.getTopics())
+        .containsExactly(singletonList(null), List.of(LogTopic.fromHexString(TOPIC_THREE)));
+  }
+
+  @Test
+  public void twoTopicsInFirstAndSecondPositionDecodesCorrectly() throws java.io.IOException {
+    final FilterParameter filterParameter =
+        readJsonAsFilterParameter(
+            "{\"topics\":["
```
Comment (rain-on) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: nit: is it overkill to use a StringJoiner to make this pretty?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java [hunk 4]
```diff
@@ -69,13 +77,15 @@ public void jsonWithSingleAddressShouldSerializeSuccessfully() throws Exception
   @Test
   public void jsonWithSingleAddressAndSingleTopicShouldSerializeSuccessfully() throws Exception {
     final String jsonWithSingleAddress =
-        "{\"jsonrpc\":\"2.0\",\"method\":\"eth_getLogs\",\"params\":[{\"address\":\"0x0\", \"topics\":\"0x0000000000000000000000000000000000000000000000000000000000000002\" }],\"id\":1}";
+        "{\"jsonrpc\":\"2.0\",\"method\":\"eth_getLogs\",\"params\":[{\"address\":\"0x0\", "
```
[the remark this answers is not in the document]
Comment (rolfyone) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: so i only touched this string to remove the duplication of 0x02...

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java (hunk 3 above)
Comment (rolfyone) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: so my original problem was that I couldn't read the tests.... this makes it pretty explicit, I can simplify it down and reduce that if you like...

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rain-on) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: 👍

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java (hunk 3 above)
Comment (rain-on) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: new StringJoiner("[", "\",\", "]") might do kinda what you want? (then do a joiner.build)

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java [hunk 5]
```diff
@@ -224,6 +238,83 @@ public void emptyListAsSubTopicDecodesCorrectly() throws JsonProcessingException
             emptyList());
   }
 
+  @Test
+  public void emptyListOfTopicsInJsonGivesInEmptyListInFilterParameter()
```
Comment (CjHare) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: Method naming in this test class is the format _[premise] should [outcome]_. 

I'd suggest a name such as `emptyTopicsListShouldDeserializeAsEmptyList()`

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java [hunk 6]
```diff
@@ -224,6 +238,83 @@ public void emptyListAsSubTopicDecodesCorrectly() throws JsonProcessingException
...
+      throws java.io.IOException {
+    final FilterParameter filterParameter = readJsonAsFilterParameter("{\"topics\":[]}");
+
+    assertThat(filterParameter.getTopics().size()).isEqualTo(0);
+  }
+
+  @Test
+  public void emptyListOfTopicsNestedGivesEmptyListInTopics() throws java.io.IOException {
```
Comment (CjHare) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameterTest.java]: Maybe `nestedEmptyTopicsListShouldDeserializeAsEmptyList()`?

(also it is more usual for unqualified exceptions e.g. `IOException`)

**Pull request merged**